src/main/java/org/breedinginsight/brapps/importer/services/processors/experiment/middleware/Middleware.java

```diff
@@ -0,0 +1,60 @@
+package org.breedinginsight.brapps.importer.services.processors.experiment.middleware;
+
+import org.breedinginsight.brapps.importer.services.processors.experiment.model.MiddlewareError;
+
+public abstract class Middleware<T> {
+
+    Middleware next;
+    Middleware prior;
+
+    /**
+     * Builds chains of middleware objects.
+     */
+    public static Middleware link(Middleware first, Middleware... chain) {
+        Middleware head = first;
+        for (Middleware nextInChain: chain) {
+            nextInChain.prior = head.getLastLink();
+            head.getLastLink().next = nextInChain;
+            head = nextInChain;
+        }
+        return first;
+    }
+
+    /**
+     * Subclasses will implement this local transaction.
+     */
+    public abstract boolean process(T context);
+    /**
+     * Subclasses will implement this method to handle errors and possibly undo the local transaction.
+     */
+    public abstract boolean compensate(T context, MiddlewareError error);
```

Comment:
The middleware is a linked list that is flexible in how it can be used. The process/processNext methods are used to move forward in the list. When an exception is caught in a link, the compensate/compensatePrior methods are used to move back through the prior links. The body of the compensate method can contain any appropriate logic, including restoring state changes resulting from a POST or PUT. For PUTs the state restoration is implemented. For POSTs, TODO placeholders are in place as reminders to implement once the DELETE endpoints are added to the BrAPIJavaTestServer.

The entity, action, and state are all required to manage an undo operation and keeping these together in the link works fine for this first pass. However, we could easily make future improvements in how these three quantities are handled since they each belong to separate classes.